### PR TITLE
Move protocol implementations to separate packages

### PR DESCRIPTION
**Pull request by jacobmcnamee**
Separates protocol implementation from the ZMQ infrastructure. Makes it possible to add a new protocol (router, adapter, port mode setting) without touching those packages.

- ZMQ router: load config from YAML files
- ZMQ adapter: load framer / filter functions from libraries
- `piksi_system_daemon`: discover protocols from libraries

TODO:
- [x] Test

/cc @swift-nav/firmware

### CODE REVIEW

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-26-g0ab9714
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-26-g0ab9714/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-26-g0ab9714/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-26-g0ab9714

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-34-g54168c5
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-34-g54168c5/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-34-g54168c5/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-34-g54168c5

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-34-g2105f01
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-34-g2105f01/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-34-g2105f01/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-34-g2105f01

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-35-gfc3424f
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-35-gfc3424f/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-35-gfc3424f/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-35-gfc3424f

File: package/nmea_protocol/src/nmea_router.yml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+
+name: NMEA_ROUTER
+ports:
+  - name: NMEA_PORT_FIRMWARE
+    pub_addr: "@tcp://127.0.0.1:44010"
+    sub_addr: "@tcp://127.0.0.1:44011"
+    forwarding_rules:
+      - dst_port: NMEA_PORT_EXTERNAL
+        filters:
+          - { action: ACCEPT }
```
Comment (mookerji) [inline package/nmea_protocol/src/nmea_router.yml]: is there an alternative way of describing this? Like: 

```
filters:
  - action:
    name: ACCEPT
```

File: package/nmea_protocol/src/nmea_router.yml [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+
+  - name: NMEA_PORT_EXTERNAL
+    pub_addr: "@tcp://127.0.0.1:44030"
```
Comment (mookerji) [inline package/nmea_protocol/src/nmea_router.yml]: Does this need to be quoted?

File: package/rtcm3_protocol/src/rtcm3_router.yml [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+    pub_addr: "@tcp://127.0.0.1:45010"
+    sub_addr: "@tcp://127.0.0.1:45011"
+
+  - name: RTCM3_PORT_EXTERNAL
+    pub_addr: "@tcp://127.0.0.1:45030"
+    sub_addr: "@tcp://127.0.0.1:45031"
+    forwarding_rules:
+      - dst_port: RTCM3_PORT_INTERNAL
+        filters:
+          - { action: ACCEPT }
```
Comment (mookerji) [inline package/rtcm3_protocol/src/rtcm3_router.yml]: Same comment here about delimeters.

File: package/nmea_protocol/src/info_nmea.c [hunk 4]
```diff
@@ -10,24 +10,19 @@
...
-uint32_t framer_sbp_process(void *framer_sbp_state,
-                            const uint8_t *data, uint32_t data_length,
-                            const uint8_t **frame, uint32_t *frame_length);
+const char *protocol_name = "NMEA";
+const char *setting_name = "NMEA";
 
-#endif /* SWIFTNAV_FRAMER_SBP_H */
+int port_adapter_opts_get(char *buf, size_t buf_size, const char *port_name)
+{
+  return snprintf(buf, buf_size,
```
Comment (mookerji) [inline package/nmea_protocol/src/info_nmea.c]: Why is this hardcoded and not in YAML?

File: package/piksi_system_daemon/src/main.c [hunk 5]
```diff
@@ -152,68 +148,110 @@ typedef struct {
...
 static adapter_config_t tcp_server1_adapter_config = {
   .name = "tcp_server1",
   .opts = "--tcp-l 55556",
-  .mode = PORT_MODE_SBP,
+  .mode = -1,
   .pid = 0
 };
 
-int port_mode_notify(void *context)
+static adapter_config_t *adapter_configs[] = {
```
Comment (mookerji) [inline package/piksi_system_daemon/src/main.c]: Why are these hardcoded and not read in through YAML?

File: package/sbp_protocol/src/sbp_router.yml [hunk 6]
```diff
@@ -0,0 +1,112 @@
+
+name: SBP_ROUTER
+ports:
+  - name: SBP_PORT_FIRMWARE
+    pub_addr: "@tcp://127.0.0.1:43010"
+    sub_addr: "@tcp://127.0.0.1:43011"
+    forwarding_rules:
+      - dst_port: SBP_PORT_SETTINGS_DAEMON
+        filters:
```
Comment (mookerji) [inline package/sbp_protocol/src/sbp_router.yml]: What about something like this, which YAML-only (vs embedding a special DSL in the YAML):

```
filters:
  - action: ACCEPT
    prefix: [0x55, 0xAE, 0x00]
  - action: ACCEPT
    prefix: [0x55, 0xA5, 0x00]
  - action: REJECT
```

File: package/nmea_protocol/src/nmea_router.yml (hunk 2 above)
Comment (jacobmcnamee) [inline package/nmea_protocol/src/nmea_router.yml]: Probably not, but some of the YAML validators I've tried have disliked leading `@`.

File: package/nmea_protocol/src/nmea_router.yml (hunk 1 above)
Comment (jacobmcnamee) [inline package/nmea_protocol/src/nmea_router.yml]: With what goal? FYI most filters contain an action and a prefix. NMEA is pretty trivial.

File: package/sbp_protocol/src/sbp_router.yml (hunk 6 above)
Comment (jacobmcnamee) [inline package/sbp_protocol/src/sbp_router.yml]: As I understand, YAML is a superset of JSON, so the `{ }` stuff is valid. I just wanted each rule on a single line for readability.

File: package/nmea_protocol/src/info_nmea.c (hunk 4 above)
Comment (jacobmcnamee) [inline package/nmea_protocol/src/info_nmea.c]: It's in a protocol-specific library sitting next to said YAML file. The YAML parser is only in `zmq_router`.

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-38-g0a00eaf
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-38-g0a00eaf/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-38-g0a00eaf/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-38-g0a00eaf

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-39-g7511a6d
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-39-g7511a6d/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-39-g7511a6d/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-39-g7511a6d

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-39-ge8049b6
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-39-ge8049b6/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-39-ge8049b6/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-39-ge8049b6

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-39-ge8049b6
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-39-ge8049b6/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-39-ge8049b6/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-39-ge8049b6

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-59-g94572d3
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-59-g94572d3/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-59-g94572d3/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-59-g94572d3

**Comment by anonymous-travis-swiftnav** [comment]
## v1.1.0-branch-63-ge6ba39d
+ [Artifacts (HITL Dashboard Page)](http://hitl-dashboard.swiftnav.com/files/swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-63-ge6ba39d/)
+ [Artifacts (S3 / AWS Console)](https://console.aws.amazon.com/s3/home?region=us-west-2&bucket=swiftnav-artifacts-pull-requests&prefix=piksi_buildroot/v1.1.0-branch-63-ge6ba39d/)
+ s3://swiftnav-artifacts-pull-requests/piksi_buildroot/v1.1.0-branch-63-ge6ba39d

**Pull request merged**